network/gossip/libp2p/test/epochtransition_test.go

```diff
@@ -0,0 +1,216 @@
+	"github.com/stretchr/testify/assert"
+	mock2 "github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+	"github.com/onflow/flow-go/model/libp2p/message"
+	"github.com/onflow/flow-go/network/gossip/libp2p"
+	protocol "github.com/onflow/flow-go/state/protocol/mock"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type EpochTransitionTestSuite struct {
+	suite.Suite
+	ConduitWrapper
+	nets         []*libp2p.Network
+	mws          []*libp2p.Middleware
+	engines      []*MeshEngine
+	state        *protocol.ReadOnlyState
+	snapshot     *protocol.Snapshot
+	ids          flow.IdentityList
+	currentEpoch int //counter to track the current epoch
+	logger       zerolog.Logger
+}
+
+func TestEpochTransitionTestSuite(t *testing.T) {
+	if _, found := os.LookupEnv("AllNetworkTest"); !found {
+		t.Skip("skipping till discovery is updated to add and remove nodes on-demand")
+	}
+	suite.Run(t, new(EpochTransitionTestSuite))
+}
+
+func (ts *EpochTransitionTestSuite) SetupTest() {
+	nodeCount := 10
+	golog.SetAllLoggers(golog.LevelError)
+	ts.logger = log.Output(zerolog.ConsoleWriter{Out: os.Stderr}).With().Caller().Logger()
+
+	// create ids
+	ids, mws := generateIDsAndMiddlewares(ts.T(), nodeCount, ts.logger)
+	ts.ids = ids
+	ts.mws = mws
+
+	// setup state related mocks
+	ts.state = new(protocol.ReadOnlyState)
+	ts.snapshot = new(protocol.Snapshot)
+	ts.state.On("Final").Return(ts.snapshot, nil)
+
+	ts.currentEpoch = 0
+	ts.updateSnapshot(0, ts.ids)
+
+	// all nodes use the same state mock
+	states := make([]*protocol.ReadOnlyState, nodeCount)
+	for i := 0; i < nodeCount; i++ {
+		states[i] = ts.state
+	}
+
+	// create networks using the mocked state and default topology
+	nets := generateNetworks(ts.T(), ts.logger, ids, mws, 100, nil, states, false)
+	ts.nets = nets
+
+	// generate the engines
+	ts.engines = generateEngines(ts.T(), nets)
+
+	// ensure that nodes can communicate with each other
+	sendMessagesAndVerify(ts.T(), ts.ids, ts.engines, ts.Publish)
+}
+
+// TearDownTest closes the networks within a specified timeout
+func (ts *EpochTransitionTestSuite) TearDownTest() {
+	for _, net := range ts.nets {
+		select {
+		// closes the network
+		case <-net.Done():
+			continue
+		case <-time.After(3 * time.Second):
+			ts.Suite.Fail("could not stop the network")
+		}
+	}
+}
+
+// TestNewNodeAdded tests that an additional node in a new epoch get connected to other nodes and can exchange messages
+func (ts *EpochTransitionTestSuite) TestNewNodeAdded() {
+
+	// create the id, middleware and network for a new node
+	ids, mws, nets := generateIDsMiddlewaresNetworks(ts.T(), 1, ts.logger, 100, nil, []*protocol.ReadOnlyState{ts.state}, false)
+	newMiddleware := mws[0]
+
+	newIDs := append(ts.ids, ids...)
+	newNetworks := append(ts.nets, nets...)
+
+	// increment epoch
+	ts.currentEpoch = ts.currentEpoch + 1
```

Comment:
@jordanschalm thanks for those details. My understanding was the same as well but I haven't yet updated the test after my initial set of changes since the test fail as peer discovery cannot be triggered on demand right now. That is something what I am working on right now. Eventually thought, the test should do what you described - add a node from the next epoch and ensure that it is part of the network.

Having said that 😃 , I will update the test and still keep it marked as skipped.

Reply:
Does this change make sense though - https://github.com/onflow/flow-go/pull/48/files#diff-e313509e481ce199478c3001443850d7880761b01f8f93ba677fab52a2176812R497 ?